### ISSUE TITLE
Azure B2C Single Sign Out/Front channel Logout URL Supported?

### ISSUE THREAD

**Issue by tsnyder-gs1us** [AzureAD/microsoft-authentication-library-for-js#3913]
### Core Library

MSAL.js v2 (@azure/msal-browser)

### Core Library Version

2.16

### Wrapper Library

Not Applicable

### Wrapper Library Version

None

### Description

For our project, we have made the necessary changes to the Azure B2C policy to support Single Sign Out and setup the Front channel logout URL. When we log out, we can trace and see both our server based (ASP.NET) and SPA (Angular using msal.js) application front channel URL is being called. The end result the user is logged out of the ASP.NET application but does not get logged out of the SPA (Angular) application.

Has anyone gotten Single Sign out working with Azure B2C and successfully gotten msal.js to logout when the Front end logout URL is called?

### Identity Provider

Azure B2C Custom Policy

### Source

External (Customer)

**Comment by samuelkubai** [comment]
Hi @tsnyder-gs1us, you can find the documentation [here](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/lib/msal-browser/docs/logout.md#front-channel-logout) on Front-channel logout and see it in action in this [sample](https://github.com/AzureAD/microsoft-authentication-library-for-js/tree/dev/samples/msal-angular-v2-samples/angular11-sample-app). Ensure your logout page meets the requirements [here](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/lib/msal-browser/docs/logout.md#requirements-for-front-channel-logout-page) for you to have a successful logout in your SPA.

**Comment by tsnyder-gs1us** [comment]
Thanks, we tried out the angular 11 sample and running into an issue. When the configured front end URL is called when get the following error. To help debug Verbose logging has been enabled. Any insight on what is causing the issue - BrowserAuthError: interaction_in_progress.

**How we are Testing**
Note for our use case we have two applications: ASP.NET Core and Angular 11 sample. We updated the sample to use our tenant and custom Azure B2C policies. We log into the Angular 11 sample, successfully, we then via SSO log into the ASP.NET Core sample. We than logout of the ASP.NET Core sample which triggers the Front channel URL and the following error occurs. 

** From Browser console.log **
[Tue, 03 Aug 2021 21:24:45 GMT] : @azure/msal-browser@2.16.1 : Verbose - Event callback registered with id: 67e3b6f0-a156-497f-a812-cd417b67b18c
main.js:542 [Tue, 03 Aug 2021 21:24:45 GMT] : @azure/msal-browser@2.16.1 : Verbose - getAllAccounts called
main.js:749 LOGOUT
vendor.js:81659 Angular is running in development mode. Call enableProdMode() to enable production mode.
main.js:542 [Tue, 03 Aug 2021 21:24:45 GMT] : @azure/msal-browser@2.16.1 : Verbose - preflightBrowserEnvironmentCheck started
main.js:542 [Tue, 03 Aug 2021 21:24:45 GMT] : @azure/msal-browser@2.16.1 : Verbose - logoutRedirect called
main.js:542 [Tue, 03 Aug 2021 21:24:45 GMT] : @azure/msal-browser@2.16.1 : Verbose - initializeLogoutRequest called
main.js:542 [Tue, 03 Aug 2021 21:24:45 GMT] : @azure/msal-angular@2.0.2 : Verbose - MsalRedirectComponent activated
main.js:542 [Tue, 03 Aug 2021 21:24:45 GMT] : @azure/msal-browser@2.16.1 : Info - Emitting event: msal:handleRedirectStart
main.js:542 [Tue, 03 Aug 2021 21:24:45 GMT] : @azure/msal-browser@2.16.1 : Verbose - Emitting event to callback 67e3b6f0-a156-497f-a812-cd417b67b18c: msal:handleRedirectStart
main.js:542 [Tue, 03 Aug 2021 21:24:45 GMT] : @azure/msal-angular@2.0.2 : Verbose - BroadcastService - msal:handleRedirectStart results in setting inProgress to handleRedirect
main.js:542 [Tue, 03 Aug 2021 21:24:45 GMT] : @azure/msal-browser@2.16.1 : Verbose - handleRedirectPromise called
main.js:542 [Tue, 03 Aug 2021 21:24:45 GMT] : @azure/msal-browser@2.16.1 : Verbose - getAllAccounts called
main.js:542 [Tue, 03 Aug 2021 21:24:45 GMT] : @azure/msal-browser@2.16.1 : Verbose - handleRedirectPromise has been called for the first time, storing the promise
main.js:542 [Tue, 03 Aug 2021 21:24:45 GMT] : @azure/msal-browser@2.16.1 : Verbose - getRedirectResponseHash called
main.js:542 [Tue, 03 Aug 2021 21:24:45 GMT] : @azure/msal-browser@2.16.1 : Verbose - Hash does not contain known properties, returning cached hash
main.js:542 [Tue, 03 Aug 2021 21:24:45 GMT] : @azure/msal-browser@2.16.1 : Info - handleRedirectPromise did not detect a response hash as a result of a redirect. Cleaning temporary cache.
vendor.js:81659 Angular is running in development mode. Call enableProdMode() to enable production mode.
main.js:542 [Tue, 03 Aug 2021 21:24:45 GMT] : @azure/msal-browser@2.16.1 : Info - Emitting event: msal:handleRedirectEnd
main.js:542 [Tue, 03 Aug 2021 21:24:45 GMT] : @azure/msal-browser@2.16.1 : Verbose - Emitting event to callback 67e3b6f0-a156-497f-a812-cd417b67b18c: msal:handleRedirectEnd
main.js:542 [Tue, 03 Aug 2021 21:24:45 GMT] : @azure/msal-angular@2.0.2 : Verbose - BroadcastService - msal:handleRedirectEnd results in setting inProgress to none
main.js:542 [Tue, 03 Aug 2021 21:24:45 GMT] : @azure/msal-browser@2.16.1 : Verbose - getAllAccounts called
vendor.js:55107 ERROR Error: Uncaught (in promise): BrowserAuthError: interaction_in_progress: Interaction is currently in progress. Please ensure that this interaction has been completed before calling an interactive API.  For more visit: aka.ms/msaljs/browser-errors.
BrowserAuthError: interaction_in_progress: Interaction is currently in progress. Please ensure that this interaction has been completed before calling an interactive API.  For more visit: aka.ms/msaljs/browser-errors.
    at BrowserAuthError.AuthError [as constructor] (vendor.js:sourcemap:3770)
    at new BrowserAuthError (vendor.js:sourcemap:136381)
    at Function.BrowserAuthError.createInteractionInProgressError (vendor.js:sourcemap:136469)
    at PublicClientApplication.ClientApplication.initializeLogoutRequest (vendor.js:sourcemap:97599)
    at PublicClientApplication.<anonymous> (vendor.js:sourcemap:96868)
    at step (vendor.js:sourcemap:89419)
    at Object.next (vendor.js:sourcemap:89349)
    at vendor.js:sourcemap:89321
    at new ZoneAwarePromise (VM99 polyfills.js:1692)
    at __awaiter (vendor.js:sourcemap:89300)
    at resolvePromise (VM99 polyfills.js:1599)
    at new ZoneAwarePromise (VM99 polyfills.js:1694)
    at __awaiter (vendor.js:sourcemap:89300)
    at PublicClientApplication.ClientApplication.logoutRedirect (vendor.js:sourcemap:96861)
    at MsalService.logoutRedirect (vendor.js:sourcemap:20585)
    at LogoutComponent.push.n1B2.LogoutComponent.ngOnInit (VM100 main.js:752)
    at callHook (vendor.js:sourcemap:50846)
    at callHooks (vendor.js:sourcemap:50807)
    at executeInitAndCheckHooks (vendor.js:sourcemap:50743)
    at refreshView (vendor.js:sourcemap:59162)

**Comment by tsnyder-gs1us** [comment]
I got the Angular sample working by setting the system timeout values. Can you verify if we need both values and what is a expects timeout value for SLO.

      iframeHashTimeout: 8000,
      loadFrameTimeout: 8000,

**Comment by samuelkubai** [comment]
Hi @tsnyder-gs1us, thank you for reaching back out. The logout page should be a dedicated logout page, that does nothing else but perform logout. Could you describe the page for front-channel logout, as based on your description and solution suspect there might be call to get a token, either by interceptor through the http service or by another route.

**Comment by tsnyder-gs1us** [comment]
Hi @samuelkubai, we were testing the msal.js Angular sample from this github repo. The logout page looks straight forward it does a console.log and then calls the logout redirect method. I think msal.js is doing something during page load. In the logger trace you can see messages coming from msal.js before the console.log from the logout page is displayed. By adding the timeout I think we are giving the iframe doing the single logout call more time to finished.

**Comment by samuelkubai** [comment]
Hi @tsnyder-gs1us, I agree the reason the timeout is fixing the issue is that it's giving the IFrame time to complete the other msal.js calls in order to complete the single logout call and thus successfully logout. To not need the timeout you would need to create a simple page for the logout action.

**Comment by tsnyder-gs1us** [comment]
Once, we have added the logout page and related checks to our full blown Angular application. We are seeing some weird timing issues where about 50% of the time the logout gets called but the user is not being logged out.

Note: Our application is Angular 12.

Were turned on Verbose logging and can see the request just stop before finishing the logout. Here are the logs differences when successfully logging out and when not. Any idea what is going on on how we determine what is causing this issue?

**Logs When Logout is Successful**
main-es2017.86e04d0….js:1 loggerCallback: [Mon, 09 Aug 2021 14:00:07 GMT] : @azure/msal-browser@2.16.1 : Verbose - preflightBrowserEnvironmentCheck started
main-es2017.86e04d0….js:1 loggerCallback: [Mon, 09 Aug 2021 14:00:07 GMT] : @azure/msal-browser@2.16.1 : Verbose - logoutRedirect called
main-es2017.86e04d0….js:1 loggerCallback: [Mon, 09 Aug 2021 14:00:07 GMT] : @azure/msal-browser@2.16.1 : Verbose - initializeLogoutRequest called
main-es2017.86e04d0….js:1 loggerCallback: [Mon, 09 Aug 2021 14:00:07 GMT] : [b65699a2-0254-4c11-a7a2-07d62c7d7921] : @azure/msal-browser@2.16.1 : Verbose - Setting postLogoutRedirectUri to current page
main-es2017.86e04d0….js:1 loggerCallback: [Mon, 09 Aug 2021 14:00:07 GMT] : [b65699a2-0254-4c11-a7a2-07d62c7d7921] : @azure/msal-browser@2.16.1 : Verbose - initializeServerTelemetryManager called
main-es2017.86e04d0….js:1 loggerCallback: [Mon, 09 Aug 2021 14:00:07 GMT] : @azure/msal-browser@2.16.1 : Info - Emitting event: msal:logoutStart
main-es2017.86e04d0….js:1 loggerCallback: [Mon, 09 Aug 2021 14:00:07 GMT] : [b65699a2-0254-4c11-a7a2-07d62c7d7921] : @azure/msal-browser@2.16.1 : Verbose - getClientConfiguration called
main-es2017.86e04d0….js:1 loggerCallback: [Mon, 09 Aug 2021 14:00:07 GMT] : [b65699a2-0254-4c11-a7a2-07d62c7d7921] : @azure/msal-browser@2.16.1 : Verbose - getDiscoveredAuthority called
main-es2017.86e04d0….js:1 loggerCallback: [Mon, 09 Aug 2021 14:00:07 GMT] : [b65699a2-0254-4c11-a7a2-07d62c7d7921] : @azure/msal-browser@2.16.1 : Verbose - Creating discovered authority with configured authority
main-es2017.86e04d0….js:1 loggerCallback: [Mon, 09 Aug 2021 14:00:07 GMT] : [c0b8a85f-df8c-450a-a3c7-67b81bad6059] : @azure/msal-browser@2.16.1 : Verbose - Silent auth client created
praveen-slo-poc-app1.azurewebsites.net/ui/main-es2017.86e04d04c9f46db64316.js:1 loggerCallback: [Mon, 09 Aug 2021 14:00:07 GMT] : @azure/msal-browser@2.16.1 : Info - Emitting event: msal:acquireTokenFromNetworkStart
praveen-slo-poc-app1.azurewebsites.net/ui/main-es2017.86e04d04c9f46db64316.js:1 loggerCallback: [Mon, 09 Aug 2021 14:00:07 GMT] : [c0b8a85f-df8c-450a-a3c7-67b81bad6059] : @azure/msal-browser@2.16.1 : Verbose - Initializing BaseAuthRequest
praveen-slo-poc-app1.azurewebsites.net/ui/main-es2017.86e04d04c9f46db64316.js:1 loggerCallback: [Mon, 09 Aug 2021 14:00:07 GMT] : [c0b8a85f-df8c-450a-a3c7-67b81bad6059] : @azure/msal-browser@2.16.1 : Verbose - Authentication Scheme set to "Bearer" as configured in Auth request
praveen-slo-poc-app1.azurewebsites.net/ui/main-es2017.86e04d04c9f46db64316.js:1 loggerCallback: [Mon, 09 Aug 2021 14:00:07 GMT] : [c0b8a85f-df8c-450a-a3c7-67b81bad6059] : @azure/msal-browser@2.16.1 : Verbose - initializeServerTelemetryManager called
praveen-slo-poc-app1.azurewebsites.net/ui/main-es2017.86e04d04c9f46db64316.js:1 loggerCallback: [Mon, 09 Aug 2021 14:00:07 GMT] : [c0b8a85f-df8c-450a-a3c7-67b81bad6059] : @azure/msal-browser@2.16.1 : Verbose - getClientConfiguration called
praveen-slo-poc-app1.azurewebsites.net/ui/main-es2017.86e04d04c9f46db64316.js:1 loggerCallback: [Mon, 09 Aug 2021 14:00:07 GMT] : [c0b8a85f-df8c-450a-a3c7-67b81bad6059] : @azure/msal-browser@2.16.1 : Verbose - getDiscoveredAuthority called
praveen-slo-poc-app1.azurewebsites.net/ui/main-es2017.86e04d04c9f46db64316.js:1 loggerCallback: [Mon, 09 Aug 2021 14:00:07 GMT] : [c0b8a85f-df8c-450a-a3c7-67b81bad6059] : @azure/msal-browser@2.16.1 : Verbose - Creating discovered authority with request authority
praveen-slo-poc-app1.azurewebsites.net/ui/main-es2017.86e04d04c9f46db64316.js:1 loggerCallback: [Mon, 09 Aug 2021 14:00:07 GMT] : [c0b8a85f-df8c-450a-a3c7-67b81bad6059] : @azure/msal-browser@2.16.1 : Verbose - Refresh token client created
praveen-slo-poc-app1.azurewebsites.net/ui/main-es2017.86e04d04c9f46db64316.js:1 loggerCallback: [Mon, 09 Aug 2021 14:00:07 GMT] : [b65699a2-0254-4c11-a7a2-07d62c7d7921] : @azure/msal-browser@2.16.1 : Verbose - Auth code client created
praveen-slo-poc-app1.azurewebsites.net/ui/main-es2017.86e04d04c9f46db64316.js:1 loggerCallback: [Mon, 09 Aug 2021 14:00:07 GMT] : [b65699a2-0254-4c11-a7a2-07d62c7d7921] : @azure/msal-browser@2.16.1 : Verbose - Setting active account to null
praveen-slo-poc-app1.azurewebsites.net/ui/main-es2017.86e04d04c9f46db64316.js:1 loggerCallback: [Mon, 09 Aug 2021 14:00:07 GMT] : @azure/msal-browser@2.16.1 : Verbose - setActiveAccount: No account passed, active account not set
praveen-slo-poc-app1.azurewebsites.net/ui/main-es2017.86e04d04c9f46db64316.js:1 loggerCallback: [Mon, 09 Aug 2021 14:00:07 GMT] : @azure/msal-browser@2.16.1 : Info - Emitting event: msal:logoutSuccess
praveen-slo-poc-app1.azurewebsites.net/ui/main-es2017.86e04d04c9f46db64316.js:1 loggerCallback: [Mon, 09 Aug 2021 14:00:07 GMT] : [b65699a2-0254-4c11-a7a2-07d62c7d7921] : @azure/msal-browser@2.16.1 : Verbose - Logout onRedirectNavigate returned false, stopping navigation
praveen-slo-poc-app1.azurewebsites.net/ui/main-es2017.86e04d04c9f46db64316.js:1 loggerCallback: [Mon, 09 Aug 2021 14:00:07 GMT] : @azure/msal-browser@2.16.1 : Info - Emitting event: msal:logoutEnd

**Logs with user is not logged out**
loggerCallback: [Mon, 09 Aug 2021 17:48:24 GMT] : @azure/msal-browser@2.16.1 : Verbose - preflightBrowserEnvironmentCheck started
main-es2017.7056d25….js:1 loggerCallback: [Mon, 09 Aug 2021 17:48:24 GMT] : @azure/msal-browser@2.16.1 : Verbose - logoutRedirect called
main-es2017.7056d25….js:1 loggerCallback: [Mon, 09 Aug 2021 17:48:24 GMT] : @azure/msal-browser@2.16.1 : Verbose - initializeLogoutRequest called
main-es2017.7056d25….js:1 loggerCallback: [Mon, 09 Aug 2021 17:48:24 GMT] : [a1550cfd-9a84-4861-a95c-d5add0cd2465] : @azure/msal-browser@2.16.1 : Verbose - Setting postLogoutRedirectUri to current page
main-es2017.7056d25….js:1 loggerCallback: [Mon, 09 Aug 2021 17:48:24 GMT] : [a1550cfd-9a84-4861-a95c-d5add0cd2465] : @azure/msal-browser@2.16.1 : Verbose - initializeServerTelemetryManager called
main-es2017.7056d25….js:1 loggerCallback: [Mon, 09 Aug 2021 17:48:24 GMT] : @azure/msal-browser@2.16.1 : Info - Emitting event: msal:logoutStart
main-es2017.7056d25….js:1 loggerCallback: [Mon, 09 Aug 2021 17:48:24 GMT] : [a1550cfd-9a84-4861-a95c-d5add0cd2465] : @azure/msal-browser@2.16.1 : Verbose - getClientConfiguration called
main-es2017.7056d25….js:1 loggerCallback: [Mon, 09 Aug 2021 17:48:24 GMT] : [a1550cfd-9a84-4861-a95c-d5add0cd2465] : @azure/msal-browser@2.16.1 : Verbose - getDiscoveredAuthority called
main-es2017.7056d25….js:1 loggerCallback: [Mon, 09 Aug 2021 17:48:24 GMT] : [a1550cfd-9a84-4861-a95c-d5add0cd2465] : @azure/msal-browser@2.16.1 : Verbose - Creating discovered authority with configured authority

**Comment by tsnyder-gs1us** [comment]
Some additional details. We see consistent behavior in Chrome and Edge. In Firefox we get an error back from B2C about trying to logout. The normal app logout works fine, we only see this error in Firefox when B2C triggers the single sign out flow.

loggerCallback: [Mon, 09 Aug 2021 20:00:47 GMT] : @azure/msal-browser@2.16.1 : Info - Emitting event: msal:logoutFailure main-es2017.e13715c413b62781c7ef.js:1:2321504
ERROR Error: Uncaught (in promise): ClientAuthError: endpoints_resolution_error: Error: could not resolve endpoints.

**Comment by tnorling** [comment]
@tsnyder-gs1us Can you please share your msal config and email me a network trace of the failure? Endpoints_resolution_error means something happened that prevented msal from retrieving the end session endpoint. This can happen for a number of reasons and if you can catch this error and inspect the `errorMessage` property that may provide additional context about why it failed in this case.

**Comment by tnorling** [comment]
@tsnyder-gs1us Based on the fiddler trace you provided it appears a network call to the ./well-known/openid-configuration endpoint is being aborted (possibly due to the front channel logout iframe being closed prematurely) which is resulting in the endpoints_resolution_error. Given that this network call is unnecessary in the front channel logout scenario I'm going to make a change to ensure this network call is not made until **after** the cache is cleared. In the meantime you can also provide the authority metadata manually to skip this network call altogether. I recommend only doing this if the page is loaded in front channel logout scenario, if possible. Directions on how to do that can be found [here](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/lib/msal-browser/docs/performance.md#bypass-authority-metadata-resolution). Let me know if that works.

**Comment by tsnyder-gs1us** [comment]
@tnorling, thanks for the suggested work around. We tested setting the authority metadata manually and the issues in Firefox has been resolved. This work around appears to resolved the timing issues we were seeing in Chrome and Edge too. We will continue testing the single logout functionality next week across all browsers.  At what point, will your PR to skip calling the authority metadata be merged into the Microsoft Authentication Library for JavaScript v2.x and released?
